src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -358,10 +359,19 @@ private <Z> Z toArray(Class<Z> arrayClass, ValueLayout elemLayout, IntFunction<Z
         return arr;
     }
 
+    public enum AccessConstraint {
+        NONE, ON_WRITE_UOE, ON_WRITE_IAE
+    }
+
     @ForceInline
-    public void checkAccess(long offset, long length, boolean readOnly) {
-        if (!readOnly && this.readOnly) {
-            throw new UnsupportedOperationException("Attempt to write a read-only segment");
+    public void checkAccess(long offset, long length, AccessConstraint access) {
```

Comment:
These changes should be reverted, I don't think we use the UOE mode anymore. Just revert the impl to what it was before (also true for VarHandle templates).